src/libshadertrap/src/executor.cc

```diff
@@ -1201,41 +1201,142 @@ bool Executor::CheckEqualBuffers(CommandAssertEqual* assert_equal) {
-    // case the contents of |mapped_buffer| cannot be null.
-    uint8_t value_1 =
-        mapped_buffer[0][index];  // NOLINT(clang-analyzer-core.NullDereference)
-    uint8_t value_2 =
-        mapped_buffer[1][index];  // NOLINT(clang-analyzer-core.NullDereference)
-    if (value_1 != value_2) {
-      std::stringstream stringstream;
-      stringstream << "Byte mismatch at index " << index << ": "
-                   << assert_equal->GetArgumentIdentifier1() << "[" << index
-                   << "] == " << static_cast<uint32_t>(value_1) << ", "
-                   << assert_equal->GetArgumentIdentifier2() << "[" << index
-                   << "] == " << static_cast<uint32_t>(value_2);
-      message_consumer_->Message(MessageConsumer::Severity::kError,
-                                 &assert_equal->GetStartToken(),
-                                 stringstream.str());
-      result = false;
+  for(auto& format_entry: format_entries){
+    switch(format_entry.kind){
+      case CommandAssertEqual::FormatEntry::Kind::kSkip:
+          start_index += format_entry.count;
+          break;
+      case CommandAssertEqual::FormatEntry::Kind::kByte:
+      {
+        for (size_t index = start_index; index < static_cast<size_t>(format_entry.count);index++) {
+          uint8_t value_1 =
+              mapped_buffer[0][index];  // NOLINT(clang-analyzer-core.NullDereference)
+          uint8_t value_2 =
+              mapped_buffer[1][index];  // NOLINT(clang-analyzer-core.NullDereference)
+          if (value_1 != value_2) {
+            std::stringstream stringstream;
+            stringstream << "Byte mismatch at index " << index << ": "
+                        << assert_equal->GetArgumentIdentifier1() << "[" << index
+                        << "] == " << static_cast<uint32_t>(value_1) << ", "
+                        << assert_equal->GetArgumentIdentifier2() << "[" << index
+                        << "] == " << static_cast<uint32_t>(value_2);
+            message_consumer_->Message(MessageConsumer::Severity::kError,
+                                      &assert_equal->GetStartToken(),
+                                      stringstream.str());
+            result = false;
+          }
+        }
+      }
+      start_index += format_entry.count;
+      break;
+      case CommandAssertEqual::FormatEntry::Kind::kFloat:
+      {
+        float* mapped_buffer_float[2]{nullptr, nullptr};
+        for (auto index : {0, 1}){
+          mapped_buffer_float[index] = reinterpret_cast<float*>(mapped_buffer[index]); 
+        }
+
+        for (size_t index = start_index; index < static_cast<size_t>(format_entry.count * sizeof(float)); index++) {
+          float value_1 =
+              mapped_buffer_float[0][index];  // NOLINT(clang-analyzer-core.NullDereference)
+          float value_2 =
+              mapped_buffer_float[1][index];  // NOLINT(clang-analyzer-core.NullDereference)
+          if (value_1 != value_2) {
```

Comment:
Clang gives a helpful warning that it's not really safe to compare floating point values via != and ==. This is something I had not thought of, but it's true: weird things can happen with NaN values, for example.

What we should do is use `std::memcmp` to compare `sizeof(float)` bytes' worth of values, and if this does not return 0 then we should read out the floating-point values and warn about them.